### DISCUSSION TITLE
OpenCV and Tensorflow Install New Version of Python

### DISCUSSION THREAD

**Discussion by amangalampalli** [Homebrew/discussions#819] [Everyday usage]
Hello all,
I was trying to install both OpenCV and Tensorflow using Homebrew onto my Intel-based MacBook Pro. However, when looking into both packages, OpenCV has not officially released support for Python3.9 and neither has Tensorflow (speculated to be upcoming in the 2.5 release). Both bottles end up downloading Python 3.9 as a dependency which I would want to avoid due to dependency issues with other libraries that only support up to Python 3.8 which Is what I have installed on my computer for the time being. Is there a way to force Homebrew to use a different installation of Python rather than downloading a completely new version? I have tried `brew edit` for both the packages in attempts to try to remove the Python 3.9 dependency and try to force them to use a version of Python I already have installed on my system, but it seems that `brew edit` seems to have no effect. Is there a way of forcing bottles to use a certain version of Python already installed on my system rather than installing a new one?

**Comment by SMillerDev** [comment]
Brew will only allow installs that it verified through CI. So bottles will only depend on the Python version specified by Homebrew.
If you do a `brew edit` you'll need to `brew install --build-from-source` to make sure the dependencies are used as in the edit.

**Reply by amangalampalli** [reply]
The stranger part of this issue is that after I edit something using `brew edit`  and then `--build-from-source`, it seems as if there is no effect as the dependencies,  paths, and modifications I have set are not even used. However, when I open the file to check the edits after the installations, my changes are still there. It seems as if Homebrew is not reading my changes even though I end up using the building from source flag.

**Reply by SMillerDev** [reply]
Did you also update the references to `Formula["python@3.9"]` when you edited them?

**Reply by amangalampalli** [reply]
I had completely removed those lines from the entire `opencv.rb` file. I had removed every occurrence of the Python3.9 formula. I have tried multiple times with just removing the Python3.9 option and also replacing them with other Python installation paths, but in the end, it had made no difference as the Python3.9 was still being downloaded.

**Reply by carlocab** [reply]
`opencv` probably still has a (possibly recursive) dependency on `python@3.9`. You can view all of them with 

```
brew deps --include-build --tree opencv
```

You'll need to completely remove `python@3.9` from `opencv`'s dependency tree to avoid it downloading `python@3.9` whenever you try to build it from source.

**Reply by amangalampalli** [reply]
Hmm, running that command showed me that almost every dependency that OpenCV had ended up having more sub dependencies that in the end almost all pointed to `python@3.9`. Is there a way I can force all these dependencies to use an installation of python that is already installed on the system without downloading a new bottle?

**Reply by carlocab** [reply]
You'll need to change all the formulae that depend on `python@3.9` so that they depend on, say, `python@3.8` instead, and then build them all from source.

You may want to `brew extract` these formulae to a tap so that your changes don't get destroyed with your next `brew {update,upgrade}`.

This is admittedly not very convenient. `brew` isn't really designed for being able to switch out dependencies easily (since dependency switches are not tested in CI).

**Reply by amangalampalli** [reply]
Unfortunately this is not very convenient but it must be done in order to use these libraries. I'm still somewhat baffled how CI was able to test for proper building of both OpenCV 4.5.1 and Tensorflow 2.4.1 considering both of them only support up to Python3.8 and Homebrew uses Python3.9 to build them. Both libraries have many module and dependency compatibility issues when using Python3.9.

**Reply by carlocab** [reply]
I guess the difficulty is that the [test](https://github.com/Homebrew/homebrew-core/blob/be8a1009f96ce3ae75f79b347bf5febdba765c38/Formula/opencv.rb#L111-L126) is trivial:

```ruby
  test do
    (testpath/"test.cpp").write <<~EOS
      #include <opencv2/opencv.hpp>
      #include <iostream>
      int main() {
        std::cout << CV_VERSION << std::endl;
        return 0;
      }
    EOS
    system ENV.cxx, "-std=c++11", "test.cpp", "-I#{include}/opencv4",
                    "-o", "test"
    assert_equal `./test`.strip, version.to_s

    output = shell_output(Formula["python@3.9"].opt_bin/"python3 -c 'import cv2; print(cv2.__version__)'")
    assert_equal version.to_s, output.chomp
  end
```

[We don't like `version`-only tests](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula), but many of them fall through the cracks. (It also looks like the `import` statement is trying to use the wrong version of `opencv`.)

If you can propose a more substantial test that actually tests the functionality of `opencv`, we might be able to look into fixing its dependencies.

**Reply by amangalampalli** [reply]
This might seem like a very strange method of testing but I remember reading that `OpenCV`, the library itself has a series of tests that compile against `Java` versions, `C++` methods, `Python2`, and `Python3` versions. What if Homebrew ended up using the library's own unit testing setup to check for the successful building of the library? Homebrew could end up triggering the specific tests needed for that particular installation (ie. if not built with `CUDA` enabled, then no need to test for `CUDA`) and the tests would also be specific to the platform (`Mac`, `Linux`, etc). The only issue I can find with this might be that it would substantially inflate the time for installation and the computer running the tests would need enough computing power. This also might work for most other libraries as well since it seems that more comprehensive tests would avoid dependency version issues such as this one in the future.

**Reply by carlocab** [reply]
No, unit tests are typically not what we look for in the `test` block. We usually aim for simulating something basic that a user might try to do with the software. Something closer to a `hello, world` test would be a much better fit.

**Reply by amangalampalli** [reply]
After a few hours of experimenting, individually changing the Python version would seem to break more and more dependencies. For small libraries/packages, this method would seem to suffice but for something as large as OpenCV, it is very tiring and tedious. There has got to be an easier way to do this, perhaps a flag to overwrite the default dependencies?

**Reply by carlocab** [reply]
No, not really, because
>`brew` isn't really designed for being able to switch out dependencies easily (since dependency switches are not tested in CI).

There is a way for you to avoid downloading `python@3.9` while installing formulae, but there's no guarantee that the formulae you install this way will use your existing Python installation. It's more likely that they just won't work because of a missing dependency.

As I mentioned earlier, if you can propose a simple test that shows that `opencv` doesn't work with `python@3.9`, then we can look into changing its dependency in homebrew/core.
